src/compiler/visitorPublic.ts

```diff
@@ -131,26 +171,43 @@ export function visitNodes<T extends Node>(nodes: NodeArray<T> | undefined, visi
     }
 
     const updated = visitArrayWorker(nodes, visitor, test, start, count);
-    if (updated !== nodes as readonly T[]) {
+    if (updated !== nodes as readonly Node[]) {
         // TODO(rbuckton): Remove dependency on `ts.factory` in favor of a provided factory.
         const updatedArray = factory.createNodeArray(updated, hasTrailingComma);
         setTextRangePosEnd(updatedArray, pos, end);
         return updatedArray;
     }
 
+    // If we are here, updated === nodes. This means that it's still a NodeArray,
+    // and also that its contents passed the tests in visitArrayWorker, so has contents
+    // of type TOut.
     return nodes;
 }
 
 /** @internal */
-export function visitArray<T extends Node, U extends T>(nodes: T[] | undefined, visitor: Visitor, test: (node: Node) => node is U, start?: number, count?: number): U[] | undefined;
-/** @internal */
-export function visitArray<T extends Node, U extends T>(nodes: readonly T[] | undefined, visitor: Visitor, test: (node: Node) => node is U, start?: number, count?: number): readonly U[] | undefined;
-/** @internal */
-export function visitArray<T extends Node>(nodes: T[] | undefined, visitor: Visitor, test: (node: Node) => node is T, start?: number, count?: number): T[] | undefined;
-/** @internal */
-export function visitArray<T extends Node>(nodes: readonly T[] | undefined, visitor: Visitor, test: (node: Node) => node is T, start?: number, count?: number): readonly T[] | undefined;
-export function visitArray<T extends Node, U extends T>(nodes: readonly T[] | undefined, visitor: Visitor, test: (node: Node) => node is U, start?: number, count?: number) {
+export function visitArray<TIn extends Node, TInArray extends readonly TIn[] | undefined, TOut extends Node>(
+    nodes: TInArray,
+    visitor: Visitor<TIn, Node | undefined> | undefined,
+    test: (node: Node) => node is TOut,
+    start?: number,
+    count?: number,
+): readonly TOut[] | (TInArray & undefined);
+export function visitArray<TIn extends Node, TInArray extends readonly TIn[] | undefined>(

```

Comment:
Why is this not marked `/** @internal */`? `visitArray` should remain internal only.

Reply:
Also a merge error. Fixing it now.